Finishing touches with regard to header improvements

alecslupu opened the pull request:
#### :tophat: What? Why?
In this PR, i am handling issues reported by #15299.

- [x] When you enter a component within a process (for example), the breadcrumb disappears and does not allow you to navigate (mobile)
- [x] `arrow_link` helper should be renamed to better reflect its new behaviour
- [x] On mobile, when I navigate to a conference program, the breadcrumb shows the [All] Conferences link instead of the Conference title. I'd expect to go back to the Conference homepage through the Conference title, since this is the only way to access the Program in the first place.
- [x]  Change the color of the disable to the one used elsewhere on the website (see reference).
- [x] All of the selectors on decidim-core/app/packs/src/decidim/dropdown_menu.js should be migrated to data selectors
- [x]  The menu is the only thing that is enabled once it is opened.


#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to #?
- Fixes #?

#### Testing
*Describe the best way to test or validate your PR.*

### :camera: Screenshots
*Please add screenshots of the changes you are proposing*
![Description](URL)

:hearts: Thank you!


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Redesigned main dropdown behavior: smoother open/close, keyboard (Escape) support, scroll locking, and improved ARIA state handling.
  * Added a subtle backdrop overlay behind dropdowns for clearer visual focus.
  * Role assignment for dropdown containers to improve assistive tech recognition.

* **Bug Fixes**
  * Mobile breadcrumb navigation now skips empty entries and falls back gracefully.

* **Tests**
  * Added comprehensive tests covering dropdown behavior, accessibility and role-assignment.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alecslupu committed:
This menu, it should be left aligned and not centered

andreslucena commented:
@alecslupu as this PR will have 18 fixes (!!!), I kindly ask you to be **really** clear in the commit messages (basically copy-pasting or rephrasing the fix from the issue) or just create a new PR for each fix. If not I'm affraid that this will be a nightmare to actual review, and you'll need to do the work two times (extracting the fixes in multiple PRs for instance once we realize that the changes are not easy to review and take too much time).

andreslucena commented:
Feel free to rewrite the current git commit history until it is ready for a review, as I didn't review it yet.

alecslupu commented:
> Feel free to rewrite the current git commit history until it is ready for a review, as I didn't review it yet.

I will do it. 
Addressing your concerns, regarding 18 fixes: Sometimes one fix can address 2 issues. 
If you want, i can extract them in their own PR, but that may impact the way you can review, as you may not see how all things fit together. 
However,  i see that 1 or 2 items from the checklist could be extracted separately ( specs, breadcrumb ), leaving only CSS issues behind. 

Allow me to get back to this (waiting for Frontend).

andreslucena commented:
>  However, i see that 1 or 2 items from the checklist could be extracted separately ( specs, breadcrumb ), leaving only CSS issues behind. 

That'd be great!

alecslupu commented:
@andreslucena This PR is the base for all the linked PRs. It is your call if we should treat the header improvements as a bundle (merge everything into this PR, then merge into develop), or as small increments (review and merge to develop independently)
- There is a lint issue (rubocop -P ) - that should be fixed by : #15848

alecslupu committed:
Merge branch 'develop' of github.com:decidim/decidim into feature/header-improvements

alecslupu committed:
Merge branch 'develop' of github.com:decidim/decidim into feature/header-improvements

alecslupu committed:
When you enter a component within a process (for example), the breadcrumb disappears and does not allow you to navigate (mobile)

alecslupu committed:
arrow_link helper should be renamed to better reflect its new behaviour

alecslupu committed:
Change the color of the disable to the one used elsewhere on the website (see reference).

alecslupu committed:
Merge branch 'develop' of github.com:decidim/decidim into feature/header-improvements

alecslupu committed:
Apply coderabbit's suggestion

alecslupu commented on decidim-core/app/views/layouts/decidim/header/_menu_breadcrumb_mobile.html.erb:
```diff
@@ -1,4 +1,4 @@
-<% previous_item = breadcrumb_items[-2] if breadcrumb_items.size > 1 %>
+<% previous_item = breadcrumb_items.reject(&:empty?)[-2] if breadcrumb_items.size > 1 %>
```
Fixed in cb0628050c

andreslucena commented:
Related to #15299

can you clean-up the body to know what it should be in this PR please? Thanks

alecslupu committed:
Fix header menu

alecslupu committed:
Apply coderabbit recommendations

andreslucena commented:
@alecslupu please be consistent in the usage of checkboxes. What does this mean?? 

<img width="1013" height="534" alt="image" src="https://github.com/user-attachments/assets/7e07b3db-52cf-400d-822d-add3b06d6ac4" />

Is only the first one done? But then I see that in this description we have 

<img width="1006" height="527" alt="image" src="https://github.com/user-attachments/assets/2118338a-067d-4115-8681-6e2731bf1e9e" />

Why one is checked and the other not? 

And in the actual PR:

<img width="1749" height="698" alt="image" src="https://github.com/user-attachments/assets/e13b9a05-f223-4439-828e-6d1d7bd18d6e" />

The file is deleted -- I thought you say that you've migrated to data selectors??? Is it intentional the removal?? WHY??? Please DO use the description and comments to communicate your intentions with the code that you're presenting: if I'm playing detective and second guessing and asking you through GH and Telegram to update the description is for you to actual take 5 minutes and communicate why you're doing these changes. 

For instance, you're linking this PR in the description: 

<img width="1018" height="489" alt="image" src="https://github.com/user-attachments/assets/c992f44f-df83-4ab2-95d8-f5c87edbf771" />

Why?? I had to double check if I was clicking correctly. 

Another thing, related to https://github.com/decidim/decidim/issues/15299 - this PR should fix all the things from there? If I go to all the comments that were pending from that issue I should see everything reflected here?

alecslupu commented on decidim-core/app/packs/src/decidim/dropdown_menu.js:
I am removing this, as i am not migrating to data selectors, as you requested, but i am moving to stimulus controllers.

andreslucena replied:
I'm ok with migrating to stimulus controllers, but I think we should still move the selectors from IDs to data attributes, to be consistent and to comply with our own conventions (and our documentation regarding these conventions)

alecslupu commented:
@andreslucena , i started working on this PR to fix all the items commented out in https://github.com/decidim/decidim/issues/15299, however, once the PR Grew bigger, i started to extract smaller prs, like breadcrumbs, mobile stuff or cleanup. 

Regarding the file removal in this Pr, is because i am moving the elements to stimulus controllers. 

As stated in our maintainer's meeting i will need to read again #15299 and start checking stuff that has already been fixed. Sometimes, an issue has been reported 2 or more times, or is related, or the fix could be a byproduct of refactor.

alecslupu commented on decidim-core/app/packs/src/decidim/controllers/header_account_menu/controller.js:
```diff
@@ -0,0 +1,12 @@
+    setTimeout(() => {
+      this.element.setAttribute("role", "dialog")
+      dropdownMobileDiv.setAttribute("role", "dialog")
+    }, 300)
```
FIxed in [5fefb72](https://github.com/decidim/decidim/pull/15683/commits/5fefb725c1226129b75d9adc4c3c6101a6e7de30)

andreslucena commented:
This wasn't done: 

https://github.com/decidim/decidim/issues/15299#issuecomment-3546211299

andreslucena commented on decidim-core/app/packs/src/decidim/controllers/header_account_menu/controller.js:
```diff
@@ -0,0 +1,18 @@
+
+export default class extends Controller {
+  connect() {
+    const dropdownMobileDiv = document.querySelector("#dropdown-menu-account-mobile");
```
This should be a data selector

andreslucena commented on decidim-core/app/packs/src/decidim/controllers/header_account_menu/controller.js:
```diff
@@ -0,0 +1,18 @@
+import { Controller } from "@hotwired/stimulus"
+
+export default class extends Controller {
```
Why do we need this class? I checked out the git commti history and you've left the most important comment: 

// changes the value "menu" of role attribute set by a11y on div dropdown-menu-account and
// dropdown-menu-account-mobile which are inappropriate for accessibility

andreslucena commented on decidim-core/app/packs/src/decidim/controllers/header_account_menu/controller.js:
```diff
@@ -0,0 +1,18 @@
+import { Controller } from "@hotwired/stimulus"
+
```
The lack of jsdoc should be an offense at the linter level. All the JS classes (specially the new ones) should have a couple of lines with the documentation/reason why they need to be.

andreslucena commented on decidim-core/app/packs/src/decidim/controllers/header_account_mobile/controller.js:
Same comments here (why this class needs to exists, data selector usage, etc).

alecslupu replied:
fixed in 9dbddaf

andreslucena replied:
Still seeing getElementById at 

    this.dropdownMobileDiv = document.getElementById(this.element.dataset.target);

alecslupu replied:
Yes, that is true. In `decidim-core/app/views/layouts/decidim/header/_main_links_mobile_item_account.html.erb` the button has 2 controllers. One dropdown, which targets the visible part using getElementById. Does not make sense to add an additional data attribute, if that is available.

Also, in the mobile menu, the elements targeted are outside the controller area, which may do the querying very slow.

andreslucena replied:
Ok. As it is already referenced through the data attributes already, I think the general idea of this rule in the conventions document 

> Do not use CSS classes to bind JavaScript selectors or Ruby specs. Try to use an id (with the js- prefix) or a data-attribute as selector, in order to shrink the coupling between changing technologies.

Should already be respected, as when refactoring/changing technologies we should detected it.

andreslucena commented on decidim-core/app/packs/src/decidim/controllers/main_menu/controller.js:
Dito

alecslupu replied:
Added jsdoc in df0af467240ada56fff38d9253d6349cf32f8722

andreslucena replied:
Still seeing the getElementById

alecslupu replied:
Yes, because the targeted element is outside the controller. 
The parent element contains too many things (login button / profile button) in mobile and desktop cases.

andreslucena replied:
Ok, as I mentioned in the other comment, if the selector is already in a data attribute I guess it should be detected in the future when making big changes in these sections

andreslucena commented on decidim-core/app/presenters/decidim/breadcrumb_root_menu_item_presenter.rb:
```diff
@@ -9,14 +9,14 @@ class BreadcrumbRootMenuItemPresenter < MenuItemPresenter
     def render
       content_tag :li, class: link_wrapper_classes do
-        output = [arrow_link(label, url, link_options)]
+        output = [participatory_space_link(label, url, link_options)]
```
I understand that this is also used for other pages that aren't spaces (such as Home), so I think we should be a bit more agnostic here in the name. As the class name is `BreadcrumbRootMenuItemPresenter`, I think `root_link` can be a good candidate

```suggestion
        output = [root_link(label, url, link_options)]
```

andreslucena commented on decidim-core/app/presenters/decidim/breadcrumb_root_menu_item_presenter.rb:
```diff
@@ -9,14 +9,14 @@ class BreadcrumbRootMenuItemPresenter < MenuItemPresenter
     end
 
-    def arrow_link(text, url, args = {})
+    def participatory_space_link(text, url, args = {})
```
```suggestion
    def root_link(text, url, args = {})
```

andreslucena reviewed (commented):
Apart from the things that I'm already mentioning, I think this PR is adding a new (broken) behavior:

<img width="1322" height="955" alt="Image" src="https://github.com/user-attachments/assets/87d1b57e-ec43-4643-adf7-752085f055ad" />


For reproducing: 

- click in the desktop menu 
- scroll
- see the page broken

andreslucena commented:
Another broken scroll behavior:

1. In mobile, open the menu dropdown-menu-main-mobile 
2. scroll
3. see that you can indeed scroll and the menu keeps opened
<img width="717" height="621" alt="image" src="https://github.com/user-attachments/assets/b3224791-688e-4900-ab3c-b135c1ce95c8" />

alecslupu committed:
Apply suggestions from code review

Co-authored-by: Andrés Pereira de Lucena <andreslucena@users.noreply.github.com>

alecslupu committed:
Fix overflow issue

alecslupu committed:
Add jsDoc for header account menu

alecslupu commented:
> Apart from the things that I'm already mentioning, I think this PR is adding a new (broken) behavior:

Explicitly this menu behavior is fixed by df0af467240ada56fff38d9253d6349cf32f8722

alecslupu commented:
> Another broken scroll behavior:
> 
>     1. In mobile, open the menu dropdown-menu-main-mobile
> 
>     2. scroll
> 
>     3. see that you can indeed scroll and the menu keeps opened
> 
> 
> <img alt="image" width="717" height="621" src="https://private-user-images.githubusercontent.com/717367/544460965-b3224791-688e-4900-ab3c-b135c1ce95c8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAxNDU2NjQsIm5iZiI6MTc3MDE0NTM2NCwicGF0aCI6Ii83MTczNjcvNTQ0NDYwOTY1LWIzMjI0NzkxLTY4OGUtNDkwMC1hYjNjLWIxMzVjMWNlOTVjOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIwM1QxOTAyNDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MjE2NTVlYWNkMTE4NGJlZDlhNjFiYzMxMTYwOTE3ZTc1ZmUyNjZlZGFlZDQ4YjRmN2E1YzI2MDBmZWU2YzBmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.U_eGCg1U-g-uRVKLmxkSODLdm8r5SeIk5SAbyZedV6c">

Fixed in f60bbaa88d23b87e86a8303de6e376b797804a4e

alecslupu committed:
Apply review recommendations

alecslupu committed:
Apply coderabbit recommendations

alecslupu commented on decidim-core/app/packs/src/decidim/controllers/assign_role/controller.js:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * This controller is used to set a role attribute for any element where is being assigned.
+ * It requires a data-role attribute with the value of the role attribute to be set.
+ * We are using to change the value "menu" of role attribute set by a11y on div dropdown-menu-account and
+ * dropdown-menu-account-mobile which are inappropriate for accessibility
+ */
+export default class extends Controller {
+  connect() {
+
+    this.timeoutId = setTimeout(() => {
+      this.element.setAttribute("role", this.element.dataset.role)
+    }, 300)
+  }
```
fixed in 9dbddaf6b3

alecslupu commented on decidim-core/app/packs/src/decidim/controllers/main_menu/controller.js:
```diff
@@ -0,0 +1,91 @@
+    this.menuButton.addEventListener("click", this.handleButtonClick)
+    document.addEventListener("keydown", this.handleKeydown)
+    if (this.closeButton) {
+      this.closeButton.addEventListener("click", this.handleDocumentClick)
+    }
+  }
+
+  disconnect() {
+    if (!this.menuContainer) {
+      return;
+    }
+
+    this.menuButton.removeEventListener("click", this.handleButtonClick)
+    document.removeEventListener("keydown", this.handleKeydown)
+    if (this.closeButton) {
+      this.closeButton.removeEventListener("click", this.handleDocumentClick)
+    }
+  }
```
Fixed in 9dbddaf6b3d0f014a899025b095eaa8356097c72

alecslupu committed:
Fix linting errors

alecslupu commented:
> This wasn't done:
> 
> [#15299 (comment)](https://github.com/decidim/decidim/issues/15299#issuecomment-3546211299)

I am pretty sure this is done back in #15737 

The reviewed code was : 
<img width="855" height="366" alt="image" src="https://github.com/user-attachments/assets/c99c180e-6774-4d49-8c8a-172bcfae1714" />

And the implementation was done in https://github.com/decidim/decidim/commit/3eb031d6575f675dc8cbbbbec3313e6e6cfc59fb

alecslupu committed:
Apply latest coderabbit suggestions

alecslupu committed:
Add js specs

alecslupu commented on decidim-core/app/packs/src/decidim/controllers/assign_role/assign_role.test.js:
```diff
@@ -0,0 +1,80 @@
+      expect(element.getAttribute("role")).toBe("navigation");
+    });
+
+    it("returns early when no data-role is provided", () => {
+      controller.disconnect();
+
+      element.dataset.role = "";
+      const setTimeoutSpy = jest.spyOn(global, "setTimeout");
+      jest.useFakeTimers();
+
+      controller.connect();
+
+      expect(element.getAttribute("role")).toBeNull();
+      expect(setTimeoutSpy).not.toHaveBeenCalled();
+      jest.advanceTimersByTime(300);
+      expect(element.getAttribute("role")).toBeNull();
+    });
```
Applying your suggestion is actually making the test to fail with the following error: 
```
    thrown: "Exceeded timeout of 5000 ms for a hook.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."
```

alecslupu commented:
@andreslucena , let me know if the `getElementById` answers are satisfactory. This should be ready for another round

andreslucena commented:
There's still a scroll weird behavior in this other menu:

<img width="463" height="869" alt="image" src="https://github.com/user-attachments/assets/fd4b1d73-ca07-4626-9f91-253089d1ebbe" />

Can you fix it too please? Thanks!

andreslucena commented:
When the menu is opened in desktop, clicking outside of it doesn't get it closed:

<img width="1850" height="1018" alt="image" src="https://github.com/user-attachments/assets/8a6b0e88-04ca-4acd-9247-b57cc30a561b" />

alecslupu committed:
Fix overflow and click on mobile menu

alecslupu committed:
Merge branch 'develop' of github.com:decidim/decidim into feature/header-improvements

alecslupu committed:
Running erblint

alecslupu commented on decidim-core/app/packs/stylesheets/decidim/_header.scss:
```diff
@@ -128,7 +128,7 @@ header {
       }
 
       &__dropdown {
-        @apply flex items-end absolute top-0 right-0 z-30 h-screen;
+        @apply flex items-end absolute top-0 right-0 bg-[rgba(0,0,0,0.25)] transition duration-300 z-30 h-screen cursor-auto;
 
```
Actually, @andreslucena requested in a comment https://github.com/decidim/decidim/pull/15683#issuecomment-3853166019, that when we click on the backdrop, the menu needs to close. therefore, we need to make sure that we have another place to click. I consider that i do not need to apply your suggestion.

alecslupu commented:
@andreslucena this should be ready for another round ...

andreslucena commented:
Merging with the failure of codecov
